src/modules/camera_trigger/camera_trigger.cpp

```diff
@@ -0,0 +1,240 @@
+#include <systemlib/param/param.h>
+#include <uORB/uORB.h>
+#include <uORB/topics/camera_trigger.h>
+#include <uORB/topics/sensor_combined.h>
+#include <uORB/topics/vehicle_command.h>
+#include <poll.h>
+#include <drivers/drv_gpio.h>
+#include <drivers/drv_hrt.h>
+#include <mavlink/mavlink_log.h>
+
+extern "C" __EXPORT int camera_trigger_main(int argc, char *argv[]);
+
+class CameraTrigger
+{
+public:
+	/**
+	 * Constructor
+	 */
+	CameraTrigger();
+
+	/**
+	 * Destructor, also kills task.
+	 */
+	~CameraTrigger();
+
+	/**
+	 * Start the task.
+	 */
+	void		start();
+
+	/**
+	 * Stop the task.
+	 */
+	void		stop();
+
+	/**
+	 * Display info.
+	 */
+	void		info();
+		
+	int 		pin;
+	
+private:
+	
+	struct hrt_call		_firecall;
+	
+	int 			_gpio_fd;
+	
+	bool	 		_trigger_enabled;
+
+	/**
+	 * Fires trigger
+	 */
+	static void	engage(void *arg);
+
+
+};
+
+namespace camera_trigger
+{
+
+CameraTrigger	*g_camera_trigger;
+}
+
+CameraTrigger::CameraTrigger() :
+	pin(1),
+	_gpio_fd(-1),
+	_trigger_enabled(true),
+	_trigger{}
+{
+	memset(&_firecall, 0, sizeof(_firecall));
+}
+
+CameraTrigger::~CameraTrigger()
+{
+	camera_trigger::g_camera_trigger = nullptr;
+}
+
+void
+CameraTrigger::start()
+{
+
+	_gpio_fd = open(PX4FMU_DEVICE_PATH, 0);
+
+	if (_gpio_fd < 0) {
+		
+		warnx("GPIO device open fail");	
+		stop();
+	}
+	else
+	{
+		warnx("GPIO device opened");
+	}
+
+
+	ioctl(_gpio_fd, GPIO_SET_OUTPUT, pin);
+
+	hrt_call_every(&_firecall, 0, 2000000, (hrt_callout)&CameraTrigger::engage, this);
+}
+
+void
+CameraTrigger::stop()
+{
+	hrt_cancel(&_firecall);
+	delete camera_trigger::g_camera_trigger;
+}
+
+void
+CameraTrigger::engage(void *arg)
+{
+
+	CameraTrigger *trig = reinterpret_cast<CameraTrigger *>(arg);
+	
+	if(trig->_trigger_enabled){
+		trig->_trigger_enabled = false; 
+		ioctl(trig->_gpio_fd, GPIO_CLEAR, trig->pin);    
+	}
+	else if(!trig->_trigger_enabled){
+		trig->_trigger_enabled = true; 
+		ioctl(trig->_gpio_fd, GPIO_SET, trig->pin);    
+	}
+}
+
+void
+CameraTrigger::info()
+{
+	warnx("Trigger state : %s", _trigger_enabled ? "enabled" : "disabled");
+}
+
+static void usage()
+{
+	errx(1, "usage: camera_trigger {start|stop|info} [-p <n>]\n"
+		     "\t-p <n>\tUse specified AUX OUT pin number (default: 1)"
+		    );
+}
+
+int camera_trigger_main(int argc, char *argv[])
+{
+	if (argc < 2) {
+		usage();
+	}
+
+	if (!strcmp(argv[1], "start")) {
+
+		if (camera_trigger::g_camera_trigger != nullptr) {
+			errx(0, "already running");
+		}
+			
+		camera_trigger::g_camera_trigger = new CameraTrigger;
+
+		if (camera_trigger::g_camera_trigger == nullptr) {
+			errx(1, "alloc failed");
+		}
+		
+		if (argc > 3) {
+	
+			camera_trigger::g_camera_trigger->pin = (int)argv[3];
+			if (atoi(argv[3]) > 0 && atoi(argv[3]) < 6) {	
+				warnx("starting trigger on pin : %li ", atoi(argv[3]));	
+				camera_trigger::g_camera_trigger->pin = atoi(argv[3]);
+			}
+			else
+			{
+				usage(); 
+			}
+		}
+		camera_trigger::g_camera_trigger->start();
```

Comment:
The start routine has to be launched by the HRT callback, not by the command line handler.